core/src/main/java/org/gbif/pipelines/core/parsers/location/LocationParser.java

```diff
@@ -0,0 +1,197 @@
+
+    // get the final country from the 2 previous parsings.
+    Country countryMatched = getFinalCountryMatch(countryName, countryCode);
+
+    // parse coordinates
+    ParsedField<LatLng> coordsParsed = parseLatLng(extendedRecord);
+
+    // add issues from coordinates parsing
+    issues.addAll(coordsParsed.getIssues());
+
+    // return if coordinates parsing failed
+    if (!coordsParsed.isSuccessful()) {
+      LOG.info("coordinates parsing failed");
+      ParsedLocation parsedLocation = ParsedLocation.newBuilder().country(countryMatched).build();
+      return ParsedField.fail(parsedLocation, issues);
+    }
+
+    // set current parsed valuies
+    ParsedLocation parsedLocation =
+      ParsedLocation.newBuilder().country(countryMatched).latLng(coordsParsed.getResult()).build();
+
+    // if the coords parsing was succesfull we try to do a country match with the coordinates
+    ParsedField<ParsedLocation> match =
+      LocationMatcher.newMatcher(parsedLocation.getLatLng(), parsedLocation.getCountry())
+        .addAdditionalTransform(CoordinatesFunction.PRESUMED_NEGATED_LAT)
+        .addAdditionalTransform(CoordinatesFunction.PRESUMED_NEGATED_LNG)
+        .addAdditionalTransform(CoordinatesFunction.PRESUMED_NEGATED_COORDS)
+        .addAdditionalTransform(CoordinatesFunction.PRESUMED_SWAPPED_COORDS)
+        .applyMatch();
+
+    // collect issues from the match
+    issues.addAll(match.getIssues());
+
+    if (match.isSuccessful()) {
+      // if match succeed we take it as result
+      parsedLocation = match.getResult();
+    }
+
+    // if the match succeed we use it as a result
+    return ParsedField.<ParsedLocation>newBuilder().successful(isParsingSuccessful(countryMatched, match))
+      .result(parsedLocation)
+      .issues(issues)
+      .build();
+  }
+
+  private static Optional<Country> getCountryParsedAndCollectIssues(
+    ParsedField<Country> countryParsed, List<InterpretationIssue> issues
+  ) {
+    if (!countryParsed.isSuccessful()) {
+      LOG.info("country parsing failed");
+      issues.addAll(countryParsed.getIssues());
+    }
+
+    return Optional.ofNullable(countryParsed.getResult());
+  }
+
+  private static Country getFinalCountryMatch(Optional<Country> countryName, Optional<Country> countryCode) {
+    // We take the country code parsing as default
+    if (countryCode.isPresent()) {
+      return countryCode.get();
+    }
+    if (countryName.isPresent()) {
+      return countryName.get();
+    }
+
+    return null;
+  }
+
+  private static boolean isParsingSuccessful(Country countryMatched, ParsedField<ParsedLocation> match) {
+    return match.isSuccessful() && countryMatched != null;
+  }
+
+  private static void checkCountryMismatch(
+    List<InterpretationIssue> issues, Optional<Country> countryName, Optional<Country> countryCode
+  ) {
+    if (countryName.isPresent() && countryCode.isPresent() && !countryName.get().equals(countryCode.get())) {
```

Comment:
I think that Optional.equals() can replace this expression: countryName.get().equals(countryCode.get())